### PR TITLE
Fix Progesss tab scrolling performance for Lesson and Level views

### PR DESCRIPTION
**Pull request by mvkski**
Scrolling through data in the Lesson and Level views under the Progress tab was
very slow for large sections. This was because the graphing dependancy we use,
react-virtualize multigrid, has a performance limitation around de-rendering
and re-rending rows if the number of rows is large and each row takes a long
time to render, if using multigrid's native functionality to only render a
subsection of a grid at a time.

This diff works around this issue by having react-virtualize multigrid render
all rows for our tables at once, rather than only rendering only a subsection
of rows at a time. This removes the need to de-render and re-render rows, and
thus removes our performance problem

Scrolling before change:
mega_levels_old
mega_lessons_old
Scrolling after change:
mega_progress
small_levels

## Links

- [jira](https://codedotorg.atlassian.net/browse/LP-764)

## Testing story

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Comment by jmkulwik** [comment]
Question from looking at the new UI:
It looks like the header scrolls off the page when the section size is large. Is it possible to keep the header stationary? Asking because with a lot of data like this, it becomes easy to lose track of where you are in the grid.

~Edit: I think I understand why it's not possible with this approach. Better question: Is this something we're ok with from a product/user standpoint?~ - Asked as a follow-up

**Comment by mvkski** [comment]
> Question from looking at the new UI:
> It looks like the header scrolls off the page when the section size is large. Is it possible to keep the header stationary? Asking because with a lot of data like this, it becomes easy to lose track of where you are in the grid.

That sounds like a good idea. My prefference would be to include this in a future PR, since I think what we have here is still a strict improvement on what we had.

File: apps/src/templates/sectionProgress/detail/VirtualizedDetailView.jsx
Comment (jmkulwik) [inline apps/src/templates/sectionProgress/detail/VirtualizedDetailView.jsx:253]: Would we need to make a similar change to improve performance for horizontal scrolling?

**Comment by jmkulwik** [comment]
> > Question from looking at the new UI:
> > It looks like the header scrolls off the page when the section size is large. Is it possible to keep the header stationary? Asking because with a lot of data like this, it becomes easy to lose track of where you are in the grid.
> 
> That sounds like a good idea. My prefference would be to include this in a future PR, since I think what we have here is still a strict improvement on what we had.

Follow-up: Is the header disappearing something we're ok with from a product/user standpoint?

File: apps/src/templates/sectionProgress/detail/VirtualizedDetailView.jsx
Comment (mvkski) [reply apps/src/templates/sectionProgress/detail/VirtualizedDetailView.jsx:253]: No such change is required, there are no horizontal scrolling performance problems at this time

**Comment by cforkish** [comment]
I agree with doing the pinned header as a follow-up task since this does seem like a huge improvement. @mvkski can you create a story for it so we're sure to discuss putting it on the roadmap?

**Comment by mvkski** [comment]
> > > Question from looking at the new UI:
> > > It looks like the header scrolls off the page when the section size is large. Is it possible to keep the header stationary? Asking because with a lot of data like this, it becomes easy to lose track of where you are in the grid.
> > 
> > 
> > That sounds like a good idea. My prefference would be to include this in a future PR, since I think what we have here is still a strict improvement on what we had.
> 
> Follow-up: Is the header disappearing something we're ok with from a product/user standpoint?

Thats a question for Amanda when she is back in the office (I will be OOTO next week when she is back).

**Comment by mvkski** [comment]
> I agree with doing the pinned header as a follow-up task since this does seem like a huge improvement. @mvkski can you create a story for it so we're sure to discuss putting it on the roadmap?

I think it'd be a good idea to run it by Amanda first in case she has anything else to say/request before we start making more tasks for this.

**Review by cforkish — APPROVED** [review]
i love how simple the solution ended up being! LGTM!

**Comment by cforkish** [comment]
> > I agree with doing the pinned header as a follow-up task since this does seem like a huge improvement. @mvkski can you create a story for it so we're sure to discuss putting it on the roadmap?
> 
> I think it'd be a good idea to run it by Amanda first in case she has anything else to say/request before we start making more tasks for this.

definitely should run it by Amanda, but creating a JIRA task for it is a good way to make sure we don't forget about it. if Amanda doesn't want to resource it she can close/delete the task.

**Review by jmkulwik — CHANGES_REQUESTED** [review]
Sorry to use the big red x - loosing the header has the potential for push-back from teachers. I'm not convinced it ends up fixing our problem. I'd like to get signoff from Product or PL (or maybe Curriculum) before merging.

**Comment by jmkulwik** [comment]
Some more details for the "requested changes."
From what I understand, the main problem we're fixing here is that it's too difficult for teachers to easily view all their students' progress and make meaningful conclusions within a reasonable amount of time. Right now performance is what's causing that problem. However, if we allow the header to disappear, we'll once again introduce that same problem, but with a different flavor. It seems like we're trading one difficulty for another. That's why I'd like to get signoff from Product or PL or Curriculum prior to merging. Someone who's a little closer to the user should be able to determine if this is an improvement to the overall issue users are experiencing.

**Comment by mvkski** [comment]
> Sorry to use the big red x - loosing the header has the potential for push-back from teachers. I'm not convinced it ends up fixing our problem. I'd like to get signoff from Product or PL (or maybe Curriculum) before merging.

How are we losing the header? The header is still there, you just need to scroll up to see it which I personally find is pretty intuitive, even if your proposed changes provide a better UX.

Agree on having product sign off before shipping.

**Comment by jmkulwik** [comment]
> > Sorry to use the big red x - loosing the header has the potential for push-back from teachers. I'm not convinced it ends up fixing our problem. I'd like to get signoff from Product or PL (or maybe Curriculum) before merging.
> 
> How are we losing the header? The header is still there, you just need to scroll up to see it which I personally find is pretty intuitive, even if your proposed changes provide a better UX.
> 
> Agree on having product sign off before shipping.

My main concern is around grading. Teachers are often switching between tabs or programs to grade student work (and generally don't have a dual monitor setup to split the various windows). I worry that without the header always visible, it will become too difficult to keep track of which lessons they are grading. It's less of an issue in the level view, where there are numbers in the bubbles, but the ~level~ lesson view, with all uniform squares, could pose problems.

Edit: By "losing the header" I meant having the header go off screen. I agree that it's pretty intuitive to just scroll up to see it again. The concern is around having it be invisible when you're part way through a list of students.

**Comment by mvkski** [comment]
> > > Sorry to use the big red x - loosing the header has the potential for push-back from teachers. I'm not convinced it ends up fixing our problem. I'd like to get signoff from Product or PL (or maybe Curriculum) before merging.
> > 
> > 
> > How are we losing the header? The header is still there, you just need to scroll up to see it which I personally find is pretty intuitive, even if your proposed changes provide a better UX.
> > Agree on having product sign off before shipping.
> 
> My main concern is around grading. Teachers are often switching between tabs or programs to grade student work (and generally don't have a dual monitor setup to split the various windows). I worry that without the header always visible, it will become too difficult to keep track of which lessons they are grading. It's less of an issue in the level view, where there are numbers in the bubbles, but the level view, with all uniform squares, could pose problems.
> 
> Edit: By "losing the header" I meant having the header go off screen. I agree that it's pretty intuitive to just scroll up to see it again. The concern is around having it be invisible when you're part way through a list of students.

That makes sense. Thanks for providing the additional context!

**Review by jmkulwik — APPROVED** [review]
Clearing the "Request for Changes" - You're already planning to chat with product, so it doesn't make sense for me to be the one blocking this anymore.

**Pull request closed without merging**
